src/main/java/servico/bookstore/Bookstore.java

```diff
@@ -703,66 +705,78 @@ public Stock getStock(int bId) {
     }
+    
+    public void updateRelatedBooks(Book targetBook) {
+        // Identificar os clientes que compraram o livro alvo nas últimas 10.000 ordens
+        Set<Integer> clientIds = getClientIdsWhoBoughtTargetBook(targetBook);
 
-    /**
-     * For all the clients that bought this book in the last 10000 orders, what
-     * are the five most sold books except this one.
-     *
-     * @param targetBook - O livro que terá os relacionados atualizados
-     *
-     */
-    private void updateRelatedBooks(Book targetBook) {
-        HashSet<Integer> clientIds = new HashSet<>();
-        int j = 0;
-        Iterator<Order> i = ordersByCreation.iterator();
-        while (i.hasNext() && j <= 10000) {
-            Order order = i.next();
-            for (OrderLine line : order.getLines()) {
-                Book book = line.getBook();
-                if (targetBook.getId() == book.getId()) {
-                    clientIds.add(order.getCustomer().getId());
-                    break;
-                }
-            }
-            j++;
+        // Contar os livros mais vendidos por esses clientes
+        Map<Integer, BookstoreBookCounter> bookCounters = countBooksBoughtByClients(clientIds, targetBook);
+
+        // Obter os cinco livros mais vendidos, exceto o alvo
+        Book[] relatedBooks = getTopFiveRelatedBooks(bookCounters, targetBook);
+
+        // Atualizar os livros relacionados no livro alvo
+        setRelatedBooks(targetBook, relatedBooks);
+    }
+
+    private Set<Integer> getClientIdsWhoBoughtTargetBook(Book targetBook) {
+        Set<Integer> clientIds = new HashSet<>();
+        Iterator<Order> orderIterator = ordersByCreation.iterator();
+
+        // Itera sobre as últimas 10.000 ordens e encontra os clientes que compraram o livro alvo
+        for (int i = 0; orderIterator.hasNext() && i < 10000; i++) {
+            Order order = orderIterator.next();
+            order.getLines().stream()
+                .filter(line -> line.getBook().getId() == targetBook.getId())
+                .findFirst()
+                .ifPresent(line -> clientIds.add(order.getCustomer().getId()));
         }
-        HashMap<Integer, Counter> counters = new HashMap<>();
-        i = ordersByCreation.iterator();
-        while (i.hasNext()) {
-            Order order = i.next();
-            if (clientIds.contains(order.getCustomer().getId())) {
-                order.getLines().forEach((line) -> {
+
+        return clientIds;
+    }
+
+    private Map<Integer, BookstoreBookCounter> countBooksBoughtByClients(Set<Integer> clientIds, Book targetBook) {
+        Map<Integer, BookstoreBookCounter> counters = new HashMap<>();
+
+        // Para cada ordem, se o cliente comprou o livro alvo, conta os outros livros comprados
+        ordersByCreation.stream()
+            .filter(order -> clientIds.contains(order.getCustomer().getId()))
+            .forEach(order -> {
+                order.getLines().forEach(line -> {

```

Comment:
```suggestion
    **
     * Counts the frequency of books purchased by the specified customers, excluding the target book.
     *
     * @param clientIds set of customer IDs to analyze purchases for
     * @param targetBook the book to exclude from the counting
     * @return a map of book IDs to their purchase frequency counters
     */
    private Map<Integer, BookstoreBookCounter> countBooksBoughtByClients(Set<Integer> clientIds, Book targetBook) {
        Map<Integer, BookstoreBookCounter> purchaseFrequency = new HashMap<>();

        ordersByCreation.stream()
            .filter(order -> clientIds.contains(order.getCustomer().getId()))
            .forEach(order -> {
                order.getLines().forEach(line -> {
```
adicionei o java doc e mudei o nome da variavel